src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -2020,11 +2020,28 @@ internal Task<Compilation> GetCompilationAsync(Project project, CancellationToke
                 : SpecializedTasks.Default<Compilation>();
         }
 
-        internal Task<bool> HasCompleteReferencesAsync(Project project, CancellationToken cancellationToken)
+        /// <summary>
+        /// Return reference completeness for the given project.
+        /// </summary>
+        internal Task<bool> IsCompleteAsync(Project project, CancellationToken cancellationToken)
+        {
+            // return IsComplete when compilation is not supported. 
+            // regardless whether project support compilation or not, if projectInfo is not complete, we can't gurantee its reference completeness
+            return project.SupportsCompilation
+                ? this.GetCompilationTracker(project.Id).IsCompleteAsync(this, cancellationToken)
+                : project.IsComplete ? SpecializedTasks.True : SpecializedTasks.False;
```

Comment:
It seems massively strange for us to be asking "is the project complete?" and the solution then goes and doesn't query the exact property on the project saying if hte project "IsComplete" or not.

If this IsComplete is only set by the host, then maybe we call it "IsConsideredCompleteByHost" so that it's more clear.

Right now, much of this PR is confusing becaues we're using the same names to represent different concepts.  Keeping track of htings is quite difficult, and it's hard to look at a piece of code and go "ok... that makes sense why it's doing what it's doing".  Naming (and possibly appropriate comments) would go a long way to making it clear what's happening.